### PR TITLE
Add conversion from 5% bucketing scheme in binary decaying histogram

### PR DESCRIPTION
**Pull request by evansheng-db**
Adds conversion logic to convert from 5% bucketing scheme to new default (1%).

We first check for annotations on the VPA Checkpoint CR, and parse the number of buckets for each binary decaying histogram.
Since we cannot easily convert from number of buckets to growth rate (due to floating point + rounding), we only support the 5% growth rate case (176 buckets).

This PR also fixes a couple tests (oom adding to +1 bucket test), as well as the epsilon values, which were incorrectly dividing the bucket length by the actual value. Instead, we should just be checking the bucket length.

## Testing
Successfully deployed this change to `dev-azure-westus` - vpa recommender is healthy and processing VPA objects.
Additionally, number of buckets changed from 176 (5%) to 696 (1%)
image

Additionally spot checked a number of VPA objects + VPA checkpoints ([doc](https://docs.google.com/document/d/1Vph62EVfQNpOtZro9RhJmU8eEapUCcCNuAQrRJLmySI/edit?tab=t.0)) and saw that recommendations did not change (outside of bounds of bucketing scheme change error), and vpa checkpoints change as expected.

**Before**
image
image

**After**
image
image

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go [hunk 1]
```diff
@@ -284,3 +286,48 @@ func (h *binaryDecayingHistogram) LoadFromCheckpoint(checkpoint *vpa_types.Histo
...
+// LoadFromCheckpoint loads the histogram from the checkpoint.
+// It checks if a conversion on bucketing scheme is necessrary, and calls the underlying LoadFromCheckpointInternal function.
+func (h *binaryDecayingHistogram) LoadFromCheckpoint(checkpoint *vpa_types.HistogramCheckpoint) error {
+	if checkpoint == nil {
+		return fmt.Errorf("cannot load from empty checkpoint")
+	}
+	// If the checkpoint does not have the same number of buckets as the histogram, we need to convert the checkpointed histogram
+	// 176 buckets indicates 5% growth rate.
+	// We don't dynamically calculate the growth rate because from numBuckets because it is not accurate due to floating points and rounding.
+	if checkpoint.NumBuckets != h.options.NumBuckets() {
```
Comment (jodzga) [inline vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go:299]: Which code path covers the case where the checkpoint does not have the annotation?
Should we panic in this case? Or should we just assume that the growth is 5%?

**Review by jodzga — CHANGES_REQUESTED** [review]
I am not sure how will we roll it out:
1. Will we first roll out the previous change fully (the one that adds annotation to checkpoints) and thus assume that annotation is always present? In that case we should probably check for missing annotation and panic as a safety measure.
2. We will roll out fully only this version but in this case we need to cleanly handle case when annotation is missing and perhaps assume that bucket growth is 5%.

File: vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go (hunk 1 above)
Comment (evansheng-db) [reply vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go:299]: @jodzga changed code to check if numbuckets is 0 (default value). This way, if the annotation doesn't exist, we will convert from 5% to 1%

File: vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go [hunk 2]
```diff
@@ -284,3 +286,48 @@ func (h *binaryDecayingHistogram) LoadFromCheckpoint(checkpoint *vpa_types.Histo
...
+			oldH := &binaryDecayingHistogram{
+				options:       oldOptions,
+				bucketForDay:  make([]uint16, h.retentionDays),
+				lastDayIndex:  0,
+				retentionDays: h.retentionDays,
+			}
+			oldH.LoadFromCheckpointInternal(checkpoint)
+			h.convertFromDifferentHistogramBucketScheme(oldH)
+		} else if checkpoint.NumBuckets != 0 {
+			return fmt.Errorf("cannot load from checkpoint: checkpoint has different number of buckets %d than the histogram %d", checkpoint.NumBuckets, h.options.NumBuckets())
```
Comment (jodzga) [inline vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go]: Shouldn't we panic here because something went horribly wrong?

Comment (evansheng-db) [reply vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go]: yep, good call - just fixed

**Review by jodzga — APPROVED** [review]

**Pull request merged**